src/main/scala/mesosphere/util/state/zk/ZKStore.scala

```diff
@@ -0,0 +1,127 @@
+package mesosphere.util.state.zk
+
+import java.util.UUID
+
+import com.fasterxml.uuid.impl.UUIDUtil
+import com.twitter.zk.ZNode.Data
+import com.twitter.zk.{ ZkClient, ZNode }
+import mesosphere.marathon.{ StoreCommandFailedException, Protos }
+import mesosphere.util.state.{ PersistentEntity, PersistentStore }
+import com.twitter.util.{ Await, Throw, Return, Future => TWFuture }
+import ZKStore._
+import org.apache.log4j.Logger
+import com.google.protobuf.ByteString
+import org.apache.zookeeper.KeeperException
+import org.apache.zookeeper.KeeperException.NoNodeException
+
+import scala.concurrent.{ Promise, Future }
+
+class ZKStore(client: ZkClient, rootNode: ZNode) extends PersistentStore {
+
+  private[this] val log = Logger.getLogger(getClass)
+
+  val root = createNode(rootNode)
+
+  /**
+    * Fetch data and return entity.
+    * The entity is returned also if it is not found in zk, since it is needed for the store operation.
+    */
+  override def load(key: ID): Future[ZKEntity] = {
+    val node = root(key)
+    require(node.parent == root, "Nested paths are not supported!")
+    node.getData().liftToTry.map {
+      case Return(data)               => ZKEntity(node, ZKData(data.bytes), Some(data))
+      case Throw(ex: NoNodeException) => ZKEntity(node, ZKData(key, UUID.randomUUID()))
+      case Throw(ex: KeeperException) => throw new StoreCommandFailedException(s"Can not fetch $key", ex)
+      case Throw(ex)                  => throw ex
+    }.asScala
+  }
+
+  /**
+    * This will store a previously fetched entity.
+    * The entity will be either created or updated, depending on the read state.
+    * @return Some value, if the store operation is successful otherwise None
+    */
+  override def save(entity: PersistentEntity): Future[ZKEntity] = {
+    val zk = zkEntity(entity)
+    def update(nodeData: ZNode.Data): Future[ZKEntity] = {
+      val future: TWFuture[Data] = zk.node.setData(zk.data.toProto.toByteArray, nodeData.stat.getVersion)
+      mapFutureResult(future, s"Can not update entity $entity") { data =>
```

Comment:
I would map to a scala future as early as possible an then use the well known methods on the scala future.